app/src/main/kotlin/compiler/common/diagnostics/Diagnostic.kt

```diff
@@ -1,11 +1,17 @@
 package compiler.common.diagnostics
 
+import compiler.lexer.Location
+
 sealed class Diagnostic {
     abstract fun isError(): Boolean
 
-    class LexerError : Diagnostic() {
-        override fun isError(): Boolean {
-            TODO("Not yet implemented")
-        }
+    class LexerError(val start: Location, val end: Location?, val context: List<String>, val errorSegment: String) : Diagnostic() {
+
+        override fun isError(): Boolean = true
+
+        override fun toString() = StringBuilder()
+            .append("Unable to match token at location $start - ${end ?: "eof"}.\n")
+            .append("\t\t${context.joinToString("")}-->$errorSegment<---")
+            .toString()
     }
```

Comment:
Not sure about the context being a list of strings if we join them anyway. Maybe we can just make it a single string. Also not sure about the context containing the 3 last tokens. Having the context be the beginning of the line containing the error would make it more readable. We could also include the rest of the line.

Reply:
I thought about this being a line, but the user already receives the line number, so to give him more compact information I chose the last 3 tokens.

Reply:
It guess it's fine. My concern is that if the error is at the beginning of the line, then one of the tokens might be a newline and the error wouldn't look all that good.